index.js

```diff
@@ -377,7 +435,76 @@ function Physics (mcData, world) {
 
     const gravityMultiplier = (vel.y <= 0 && entity.slowFalling > 0) ? physics.slowFalling : 1
 
-    if (!entity.isInWater && !entity.isInLava) {
+    if (entity.isInWater || entity.isInLava) {
+      // Water / Lava movement
+      const lastY = pos.y
+      let acceleration = physics.liquidAcceleration
+      const inertia = entity.isInWater ? physics.waterInertia : physics.lavaInertia
+      let horizontalInertia = inertia
+
+      if (entity.isInWater) {
+        let strider = Math.min(entity.depthStrider, 3)
+        if (!entity.onGround) {
+          strider *= 0.5
+        }
+        if (strider > 0) {
+          horizontalInertia += (0.546 - horizontalInertia) * strider / 3
+          acceleration += (0.7 - acceleration) * strider / 3
+        }
+
+        if (entity.dolphinsGrace > 0) horizontalInertia = 0.96
+      }
+
+      applyHeading(entity, strafe, forward, acceleration)
+      moveEntity(entity, world, vel.x, vel.y, vel.z)
+      vel.y *= inertia
+      vel.y -= (entity.isInWater ? physics.waterGravity : physics.lavaGravity) * gravityMultiplier
+      vel.x *= horizontalInertia
+      vel.z *= horizontalInertia
+
+      if (entity.isCollidedHorizontally && doesNotCollide(world, pos.offset(vel.x, vel.y + 0.6 - pos.y + lastY, vel.z))) {
+        vel.y = physics.outOfLiquidImpulse // jump out of liquid
+      }
+    } else if (entity.elytraFlying) {
+      const {
+        pitch,
+        sinPitch,
+        cosPitch,
```

Comment:
Pretty much all mentions of cosPitch in your implementation seems to be flipped for `sqrt(lookDir.x * lookDir.x + lookDir.z * lookDir.z)` in Vanilla.

Reply:
For a normalized lookDir which ours is, `sqrt(lookDir.x * lookDir.x + lookDir.z * lookDir.z) == cosPitch`. For floating point reasons, is it better to match vanilla?

Reply:
Vanilla does not normalize lookDir

Reply:
Nevermind the above comment. Vanilla does normalize the view vector. Just didn't seem obvious since it has methods for calculating the length of the view vector.